File: sockets/java/src/ServidorBasico.java
import java.io.*;
import java.net.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Servidor básico mono-hilo que maneja un cliente a la vez
 */
public class ServidorBasico {
    private static final int PUERTO = 8080;
    
    public static void main(String[] args) {
        System.out.println("=== SERVIDOR BÁSICO MONO-HILO ===");
        System.out.println("Iniciando servidor en puerto " + PUERTO);
        
        try (ServerSocket servidor = new ServerSocket(PUERTO)) {
            System.out.println("Servidor iniciado. Esperando conexiones...");
            
            while (true) {
                // Acepta una conexión (BLOQUEA hasta que llegue un cliente)
                Socket cliente = servidor.accept();
                System.out.println("Cliente conectado desde: " + cliente.getInetAddress());
                
                // Maneja el cliente (UN SOLO CLIENTE A LA VEZ)
                manejarCliente(cliente);
            }
            
        } catch (IOException e) {
            System.err.println("Error del servidor: " + e.getMessage());
        }
    }
    
    private static void manejarCliente(Socket cliente) {
        try (
            BufferedReader entrada = new BufferedReader(
                new InputStreamReader(cliente.getInputStream()));
            PrintWriter salida = new PrintWriter(
                cliente.getOutputStream(), true)
        ) {
            String mensaje;
            salida.println("Conectado al servidor. Escribe 'salir' para desconectar.");
            
            while ((mensaje = entrada.readLine()) != null) {
                String timestamp = LocalDateTime.now().format(
                    DateTimeFormatter.ofPattern("HH:mm:ss"));
                
                System.out.println("[" + timestamp + "] Cliente: " + mensaje);
                
                if ("salir".equalsIgnoreCase(mensaje.trim())) {
                    salida.println("Desconectando...");
                    break;
                }
                
                // Echo del mensaje con timestamp
                String respuesta = "[" + timestamp + "] Servidor recibió: " + mensaje;
                salida.println(respuesta);
            }
            
        } catch (IOException e) {
            System.err.println("Error manejando cliente: " + e.getMessage());
        } finally {
            try {
                cliente.close();
                System.out.println("Cliente desconectado");
            } catch (IOException e) {
                System.err.println("Error cerrando conexión: " + e.getMessage());
            }
        }
    }
}

File: sockets/python/servidor_basico.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Servidor TCP básico que acepta conexiones secuenciales
Ejemplo simple de servidor que maneja un cliente a la vez
"""

import socket
import datetime

HOST = 'localhost'
PUERTO = 8080

def main():
    print("=== SERVIDOR BÁSICO ===")
    print(f"Iniciando servidor en {HOST}:{PUERTO}")
    
    # Crear socket del servidor
    servidor = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    
    try:
        # Permitir reutilizar la dirección
        servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        servidor.bind((HOST, PUERTO))
        servidor.listen(5)
        
        print("Servidor iniciado. Esperando conexiones...")
        cliente_num = 0
        
        while True:
            # Aceptar conexión del cliente (BLOQUEA hasta que llegue una)
            cliente_socket, direccion_cliente = servidor.accept()
            cliente_num += 1
            
            print(f"Cliente {cliente_num} conectado desde {direccion_cliente}")
            
            # Manejar el cliente
            manejar_cliente(cliente_socket, cliente_num)
            
    except KeyboardInterrupt:
        print("\nServidor interrumpido por el usuario")
    except Exception as e:
        print(f"Error del servidor: {e}")
    finally:
        servidor.close()
        print("Servidor cerrado")

def manejar_cliente(cliente_socket, numero_cliente):
    """
    Maneja la comunicación con un cliente específico
    """
    try:
        # Enviar mensaje de bienvenida
        timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        mensaje_bienvenida = f"Bienvenido cliente {numero_cliente}! Hora: {timestamp}\n"
        cliente_socket.send(mensaje_bienvenida.encode('utf-8'))
        
        # Leer datos del cliente
        while True:
            datos = cliente_socket.recv(1024)
            if not datos:
                break
                
            mensaje_recibido = datos.decode('utf-8').strip()
            print(f"Cliente {numero_cliente}: {mensaje_recibido}")
            
            # Si el cliente dice "adios", terminar conexión
            if mensaje_recibido.lower() == "adios":
                respuesta = "¡Hasta luego!\n"
                cliente_socket.send(respuesta.encode('utf-8'))
                break
            
            # Echo: devolver el mismo mensaje con timestamp
            timestamp = datetime.datetime.now().strftime("%H:%M:%S")
            respuesta = f"[{timestamp}] Echo: {mensaje_recibido}\n"
            cliente_socket.send(respuesta.encode('utf-8'))
            
    except Exception as e:
        print(f"Error manejando cliente {numero_cliente}: {e}")
    finally:
        cliente_socket.close()
        print(f"Cliente {numero_cliente} desconectado")

if __name__ == "__main__":
    main()